muacrypt/recommendation.py

```diff
@@ -1,34 +1,67 @@
 class Recommendation:
     """ Calculating recommendations for encryption """
 
-    def __init__(self, peerstates):
+    def __init__(self, peerstates, prefer_encrypt, reply_to_enc=False):
         self.peerstates = peerstates
+        self.prefer_encrypt = prefer_encrypt
+        self.reply_to_enc = reply_to_enc
 
     def ui_recommendation(self):
         # only consider first peer for now
-        peer = list(self.peerstates.values())[0]
-        return self._peer_recommendation(peer)
+        peer_recommendations = [
+            self._peer_recommendation(peer).ui_recommendation()
+            for peer in self.peerstates.values()]
+        precedence = ['disable', 'discourage', 'available', 'encrypt']
+        return next(rec for rec in precedence if rec in peer_recommendations)
 
     def target_keys(self):
-        return {addr: self._target_key(peer) for addr, peer in
+        return {addr: self._peer_recommendation(peer).target_key()
+                for addr, peer in
                 self.peerstates.items()}
 
     def _peer_recommendation(self, peer):
-        if self._target_key(peer):
-            return 'available'
-        else:
-            return 'disable'
+        return PeerRecommendation(peer, self.prefer_encrypt,
+                                  self.reply_to_enc)
+
+
+class PeerRecommendation:
+    """ Calculating recommendation for a single peer """
+
+    def __init__(self, peer, prefer_encrypt, reply_to_enc):
+        self.peer = peer
+        self.prefer_encrypt = prefer_encrypt
+        self.reply_to_enc = reply_to_enc
 
-    def _target_key(self, peer):
-        return self._public_key(peer) or self._gossip_key(peer)
+    def ui_recommendation(self):
+        pre = self._preliminary_recommendation()
+        if ((pre == 'available' or pre == 'discourage') and
+                self.reply_to_enc):
+            return 'encrypt'
+        if (pre == 'available' and
+                self.prefer_encrypt == 'mutual' and
+                self.peer.prefer_encrypt == 'mutual'):
+            return 'encrypt'
+        return pre
+
+    def target_key(self):
+        return self._public_key() or self._gossip_key()
+
+    def _preliminary_recommendation(self):
+        if self.target_key() is None:
+            return 'disable'
+        timeout = 35 * 24 * 60 * 60
+        if (self.peer.last_seen - self.peer.autocrypt_timestamp >
+                timeout):
+            return 'discourage'
+        return 'available'
 
-    def _public_key(self, peer):
-        return self._key(peer.public_keyhandle)
+    def _public_key(self):
+        return self._key(self.peer.public_keyhandle)
 
-    def _gossip_key(self, peer):
+    def _gossip_key(self):
         # gossip keyhandle is not implemented yet.
-        if hasattr(peer, 'gossip_keyhandle'):
-            return self._key(peer.gossip_keyhandle)
+        if hasattr(self.peer, 'gossip_keyhandle'):
+            return self._key(self.peer.gossip_keyhandle)
```

Comment:
IMO the API ("_public_key" and "_gossip_key") should have "keyhandle" in the name and just return keyhandles.  The ``_key`` function is superflous with the current code and should go.  It's easy to introduce if needed.  (i am usually against introducing abstractions that are not used but "might be used in the future").  Instead of the "if hasattr" you can then also just do `return getattr(self.peer, 'gossip_keyhandle', None)` etc.  Actually ``target_key`` can just directly do these two getattr's and the current `_*key` methods can all vanish then, reducing indirections and still having very short functions.